sematic/future_context.py

```diff
@@ -90,13 +113,37 @@ def context() -> SematicContext:
 
     Raises
     ------
-    RuntimeError:
+    NotInSematicFuncError:
         If this function is called outside the execution of a Sematic function.
     """
     global _current_context
     if _current_context is None:
-        raise RuntimeError(
+        raise NotInSematicFuncError(
             "context() must be called from within the execution of a Sematic function, "
             "in the root process that function was invoked from."
         )
     return _current_context
+
+
+class NotInSematicFuncError(RuntimeError):
+    pass
+
+
+# Note: this helps resolve circular imports. Otherwise we'd have:
+# future_context -> external_resource -> future_context
+# this approach cuts the first dependency in the above
+class AbstractExternalResource:
```

Comment:
How about storing `AbstractExternalResource` in its own module at `sematic.abstract_external_resource` and define in there as abstract methods all public interfaces (`activate`, `deactivate`, etc. – that is what we care about form a typing perspective) and keep the coupling with `future_context` in `ExternalResource` which we store alongside implementations in  the `sematic/external_resources` directory as is our pattern elsewhere.

Also why not rename `ExternalResource` to `Resource` makes it more generic and external seems unnecessary?

So putting it all together, we would have:
```
sematic/
    abstract_resource.py
    resources/
        resource.py
        spark_resource.py
        ray_resource.py
```